### PR TITLE
`sage.knots`: Modularization fixes (imports), `# needs`

### PR DESCRIPTION
**Pull request by mkoeppe**
Also some doctest cosmetics.

- Cherry-picked from #35095.

`sage --fixdoctests --verbose` now shows doctest tagging statistics.

### :memo: Checklist

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

File: src/sage/knots/link.py [hunk 1]
```diff
@@ -2548,33 +2617,34 @@ def reverse(self):
            sage: K3 = Knot([[5, 2, 4, 1], [3, 6, 2, 5], [1, 4, 6, 3]])
            sage: K3r = K3.reverse(); K3r.pd_code()
            [[4, 1, 5, 2], [2, 5, 3, 6], [6, 3, 1, 4]]
-           sage: K3 == K3r
+           sage: K3 == K3r                                                              # needs sage.groups
            True
```
Comment (kwankyu) [inline src/sage/knots/link.py]: Fix indent here too.

**Comment by kwankyu** [comment]
What distribution package (K) provides `sage.knots`?

knot theory heavily depends on groups. I think the distribution package should depend on the distribution package (G) that provides groups. What is (G)?

**Comment by mkoeppe** [comment]
In #35095, I have `sage.knots` in **sagemath-graphs**

File: src/sage/knots/link.py (hunk 1 above)
Comment (mkoeppe) [reply src/sage/knots/link.py]: Done in 9a4517551dc

**Comment by kwankyu** [comment]
I think `sage.knots` should be in `sagemath-standard-no-symbolics` for now. I am not an expert, of course, but I guess knot theorist would not regard their subject as part of graph theory.

**Comment by kwankyu** [comment]
@soehms ?

**Comment by mkoeppe** [comment]
Per design of the modularized distributions, **sagemath-graphs** is not a package only for "graph theory". Rather it's a package that defines graphs and recognizable generalizations thereof and contains code that directly makes use of such structures. And the most basic code in `sage.knots.link` is built on `DiGraph` (see for example the methods `Link.arcs`).

**Comment by mkoeppe** [comment]
this part applies to `sage.knots`:
> contains code that directly makes use of such structures.

**Comment by kwankyu** [comment]
Similarly `sage.knots` makes heavy use of groups, which is why we are forced to add `# needs sage.groups` to so many doctests.

Let's wait for an opinion from an expert though.

**Comment by kwankyu** [comment]
By the way, there are 39 instances of the word "graph" in `sage.knots` while there are 161 of "group".

**Comment by soehms** [comment]
> Similarly `sage.knots` makes heavy use of groups, which is why we are forced to add `# needs sage.groups` to so many doctests.
> 
> Let's wait for an opinion from an expert though.

I'm not really a knot theory expert either, but I think to know enought about the `knots` subfolder to discuss the question.

> I think sage.knots should be in sagemath-standard-no-symbolics for now. I am not an expert, of course, but I guess knot theorist would not regard their subject as part of graph theory.

That's right. According to Wikipedia knot theory is a part of [Geometric toplology](https://en.wikipedia.org/wiki/Geometric_topology). On that page it is listed as a subfield of its own, while on the [Low-dimensional topology](https://en.wikipedia.org/wiki/Low-dimensional_topology) page it is mentioned as a subfield as well.

As you already explained the connections to graph theory, group theory and polynomial rings just appear to be tool-boxes.

Maybe the problem with modularization is an inconsistency between the developer's and the user's perspective. If a user interested in knot theory or geometric topology doesn't want to install the whole of the Sage library to work with topologically relevant parts of it, it won't help him that we've split the functionality in terms of dependencies. 
 
To facilitate modularization for such use cases, we should find a way to declare meta-packages as collections of existing standard and optional packages that define parts of Sage that cover a certain mathematical area. For example, such a meta-package for knot theory would consist of the standard packages `sagemath-standard`, `sagemath__graphs`, `sagemath__groups`, `sagemath__rings...` and optional packages like `snappy`, `database_knotinfo`... 

But I'm not sure if such use cases are intended by our modularization project. If modularization only adresses developers, then I agree that we should not give preference to any of the tool-boxes (e.g. `sagemath__graphs` over `sagemath__groups`).

Other opinions, @culler , @NathanDunfield , @tscrim , @miguelmarco , @fchapoton ?

**Comment by mkoeppe** [comment]
@soehms Thanks for sharing this analysis. 

> Maybe the problem with modularization is an inconsistency between the developer's and the user's perspective. If a user interested in knot theory or geometric topology doesn't want to install the whole of the Sage library to work with topologically relevant parts of it, it won't help him that we've split the functionality in terms of dependencies.
> 
> To facilitate modularization for such use cases, we should find a way to declare meta-packages as collections of existing standard and optional packages that define parts of Sage that cover a certain mathematical area. For example, such a meta-package for knot theory [...] But I'm not sure if such use cases are intended by our modularization project. If modularization only adresses developers, then I agree that we should not give preference to any of the tool-boxes (e.g. `sagemath__graphs` over `sagemath__groups`).

It's true that my development focus in this "modularization" phase of the modularization project has been in getting the modularized distributions to ship and build in isolation. The dependencies are the hard technical constraints for this, and this is where the main technical difficulty lies. I have done most of this work already by myself (with the help of all who have supported the upstreaming of this work in their roles as PR reviewers); it's a necessary "bootstrapping" step. 

Sometimes decisions have to be taken that are arbitrary to some degree. So there is a component (knots) that has two hard runtime dependencies, groups and graphs. Well, one of the distributions has to ship it. It would be a mistake to defer the decision and keep knots for now in a catch-all, not-yet-modularized distribution until "we know better". Breaking the vicious cycle, making everything shippable in practical distributions, is exactly the point of this phase.

Defining convenient and stable ways to advertise functionality to users is a largely separate question, and this is one where I indeed hope that communities of domain experts -- developers and users -- will start to engage with the modularization project. This is really the "integration" phase of the modularization. Most of the technical difficulty is gone: defining meta-packages is just a matter of writing a pyproject.toml file; with "optional dependencies" that advertise features. So a different hard problem comes to the fore: Reaching out to users, shaping and leading communities.

**Comment by NathanDunfield** [comment]
I agree with @soehms and @kwankyu that, mathematically, knot theory is not usually viewed as part of (or even strongly connected to) graph theory.  There's more overlap with modern combinatorial/geometric group theory, especially the aspects that `sage.knots` implements.

On the technical question of where `sage.knots` should go in the modularization, I have no informed opinion.

**Comment by culler** [comment]
Here is my position on modularization:

1. Modularization is NOT for users.  A user who wants to use Sage should
install Sage.  Period.  (And the Sage project should put effort into making
it as easy as possible to install Sage.)

2. Modularization is very useful for developers who would like to include
some parts of Sage, either as a dependency or as a "vendorization" in the
project that the developer is developing.  Including all of Sage, in either
way, turns small or medium sized projects into enormous projects and
therefore is not practical.  A guiding example is CyPari, which is a
dependency of SnapPy and also is embedded in the SnapPy app.

- Marc

PS Classifying the sub-disciplines of mathematics is hard, and is not the
job of the Sage developers, and should not be a prerequisite of the
modularization project.  The decision of which Sage components to group
into a module should be based on the software.  The decision of which Sage
modules to include in a given project should be made by the project
developer based on what the project needs.

**Comment by mkoeppe** [comment]
I agree with most of what @culler wrote, except to caution that the word "user" means many different things to different people. And the transition from (1) being merely a user of software to (2) sharing a script to (3) publishing a package in the Python "ecosystem" is soft and gradual, and supporting this transition when Sage is part of the equation is one of the goals of the modularization project.

**Comment by kwankyu** [comment]
First let us note that some distribution packages are already in sage, and some other distribution packages are yet either in Matthias' repo or in future. All these collectively form the basic products of the modularization project. This PR is about preparation of the sage library for the modularization project, in particular, related with the python module `sage.knots`.

>  ... I [Matthias] have done most of this work already by myself ...

Yes. Thanks for that. However, I think this is the source of problems about the modularization project. People in the sage community did not engage in the design of the modularization of the sage library. For the project to proceed, I think it is the time when the design (and the design principles) is shared with the community, discussed and reflects the results of the discussion. Both of the users and developers of sage should acknowledge the project and become supporters of the modularized sage. Even accepting and not being hostile to `# needs` tags scattered around doctests needs support of all the sage community. 

As we are on this PR, we may take `sage.knots` as the target of the discussion.
 
In the current design, `sage.knots` is in the distribution (package) `sagemath-graphs`, according to the diagram in #35095.

This PR introduces too many `# needs sage.groups` tags in `sage.knots`, which I think signals bad design. The central question is whether `sagemath-graphs` is the right distribution  for `sage.knots`. 

According to #35095. It contains some of `sage.combinat`, most of `sage.graphs`, `sage.topology`, and `sage.knots`. The module `sage.topology` is mostly simplicial complex machinery, which uses graphs. Hence `sage.topology` is really about  algebraic topology.

A developer developing his own software may depend on `sagemath.graphs`. Then he is the developer of his software and at the same time a user of the distribution package `sagemath.graphs`.  The user expects certain functionality of `sagemath.graphs` from its name, in this case, graphs; `sage.topology` and `sage.knots` are not modules expected to be found in `sagemath-graphs`.

> ... So there is a component (knots) that has two hard runtime dependencies, groups and graphs. Well, one of the distributions has to ship it.  

Why? `sage.knots` uses groups and graphs, but does not belong to graphs and groups.

> It would be a mistake to defer the decision and keep knots for now in a catch-all, not-yet-modularized distribution until "we know better". 

If we want to ship `sage.knots` in a smaller distribution than `sagemath-standard-no-symbolics`, we may have `sagemath-topology` that contains `sage.topology` and `sage.knots` and depends on `sagemath-graphs` and `sagemath-groups`.

> Defining convenient and stable ways to advertise functionality to users is a largely separate question, and this is one where I indeed hope that communities of domain experts -- developers and users -- will start to engage with the modularization project.

The first phase will be over when the distribution packages shipping hard dependencies (the blue boxes in the diagram of   
#35095) are in sage. People should engage in the design phase of all distributions named `sagemath-mathematical-subject`. That is now.

> ... Most of the technical difficulty is gone: defining meta-packages is just a matter of writing a pyproject.toml file; with "optional dependencies" that advertise features. So a different hard problem comes to the fore: Reaching out to users, shaping and leading communities.

Having people wait until the phase of defining "meta-packages" and just accept your design as done in #35095 seems to be unrealistic strategy to proceed the modularization project, as I think the recent disputed PRs manifest.

**Comment by mkoeppe** [comment]
> Having people wait until the phase of defining "meta-packages" and just accept your design as done in #35095 seems to be unrealistic strategy to proceed the modularization project, as I think the recent disputed PRs manifest.

I disagree. These so-called "disputes" have no legitimacy whatsoever. The project will not be governed by these expressions of disrespect and hostility.

**Comment by mkoeppe** [comment]
> > ... So there is a component (knots) that has two hard runtime dependencies, groups and graphs. Well, one of the distributions has to ship it.
> 
> Why? `sage.knots` uses groups and graphs, but does not belong to graphs and groups.

I prefer not to engage with such statements of what belongs and what does not.
We have agency there as the developers.

One important criterion is the overall complexity -- how many distributions we create.

**Comment by kwankyu** [comment]
> > Having people wait until the phase of defining "meta-packages" and just accept your design as done in #35095 seems to be unrealistic strategy to proceed the modularization project, as I think the recent disputed PRs manifest.
> 
> I disagree. These so-called "disputes" have no legitimacy whatsoever. The project will not be governed by these expressions of disrespect and hostility.

Sorry that they remind you of the disrespect and hostility. Of course, it has no place in our community. I just wanted to mention that not a few people have hostile (or at least unfriendly) attitude towards the modularization itself. I think it is important for the community to embrace the modularization project as their own mission.

**Comment by soehms** [comment]
> Here is my position on modularization: 1. Modularization is NOT for users. A user who wants to use Sage should install Sage. Period. (And the Sage project should put effort into making it as easy as possible to install Sage.) 

I agree with that 99%, especially the part in parentheses 100%. On the other hand, some users may not want to spend that much disk space, so why not have an option for them?

**Comment by soehms** [comment]
> If we want to ship `sage.knots` in a smaller distribution than `sagemath-standard-no-symbolics`, we may have `sagemath-topology` that contains `sage.topology` and `sage.knots` and depends on `sagemath-graphs` and `sagemath-groups`.

From a mathematical point of view, this sounds reasonable to me, even though there are no direct dependencies on `sage.topology`. Could we add more dependencies to such a distribution package, for example `sagemath-modules` because of `sage.homology`...? Unfortunately I can't comment on the technical difficulties. But if they are limited, this would be a solution I prefer.

**Comment by kwankyu** [comment]
> From a mathematical point of view, this sounds reasonable to me, even though there are no direct dependencies on `sage.topology`. 

What do you mean by "direct dependencies on sage.topology`?

> Could we add more dependencies to such a distribution package, for example `sagemath-modules` because of `sage.homology`...? 

`sagemath-topology` would depend on `sagemath-groups` which in turn depends on `sagemath-modules`. So `sagemath-modules` becomes automatically a dependency of the supposed `sagemath-topology`. See the diagram in #35095.

By the way, for further discussions on the design of modularization, we may use github's [Discussions](https://github.com/sagemath/sage/discussions/landing) facility. If Matthias thinks the current discussion is not appropriate here in PR, we can move. I have no prior experience with github Discussions. So I don't know if it will serve us well...

**Comment by kwankyu** [comment]
> > Why? `sage.knots` uses groups and graphs, but does not belong to graphs and groups.
> 
> I prefer not to engage with such statements of what belongs and what does not. We have agency there as the developers.

I don't understand this: "We have agency there as the developers".

**Comment by soehms** [comment]
> > From a mathematical point of view, this sounds reasonable to me, even though there are no direct dependencies on `sage.topology`.
> 
> What do you mean by "direct dependencies on sage.topology`?

Just that there are no imports under `sage/knots` of this module.

> > Could we add more dependencies to such a distribution package, for example `sagemath-modules` because of `sage.homology`...?
> 
> `sagemath-topology` would depend on `sagemath-groups` which in turn depends on `sagemath-modules`. So `sagemath-modules` becomes automatically a dependency of the supposed `sagemath-topology`. See the diagram in #35095.
> 

I see! Then all the better!

> By the way, for further discussions on the design of modularization, we may use github's [Discussions](https://github.com/sagemath/sage/discussions/landing) facility. If Matthias thinks the current discussion is not appropriate here in PR, we can move. I have no prior experience with github Discussions. So I don't know if it will serve us well...

Your link gives a 404 error for me!

**Comment by kwankyu** [comment]
> > By the way, for further discussions on the design of modularization, we may use github's [Discussions](https://github.com/sagemath/sage/discussions/landing) facility. If Matthias thinks the current discussion is not appropriate here in PR, we can move. I have no prior experience with github Discussions. So I don't know if it will serve us well...
> 
> Your link gives a 404 error for me!

Sorry. The page says "Only maintainers can see this page and enable Discussions". We can use "Discussions" service if a maintainer enables it. Design of modularization seems to be a good topic for everyone to participate. I will leave this decision to Matthias the modularization project leader.

**Comment by mkoeppe** [comment]
I'd like to bring the topic back to tagging of doctests in `sage.knots`. 
After all, in this PR here, there is nothing that actually defines a distribution.

I've added file-level doctest tags `# needs sage.graphs`.

I've extended our tooling to collect doctest tagging statistics, like @NathanDunfield attempted in  https://github.com/sagemath/sage/pull/37900#issuecomment-2088914174 using just `grep`.
```
$ ./sage -fixdoctests --no-test --verbose src/sage/knots/
sage-fixdoctests: Processing unprocessed files
sage-fixdoctests: No fixes made in 'src/sage/knots/link.py'
File tags: 
       sage.graphs
Doctest blocks by persistent tags: 
    93 (untagged)
    46 sage.groups
    11 sage.groups sage.modules
     5 database_knotinfo
     2 sage.groups sage.libs.homfly
     2 database_knotinfo snappy
     1 sage.modules
     1 sage.groups sage.modules sage.rings.number_field
     1 sage.symbolic
     1 sage.libs.homfly
     1 sage.libs.pari sage.modules
     1 sage.plot
Doctest examples by tags: 
   592 (untagged)
    11 sage.groups
    11 sage.libs.homfly
    10 sage.plot
     8 sage.modules
     8 sage.symbolic
     3 sage.libs.pari sage.modules
     2 database_knotinfo
     1 sage.groups sage.modules
     1 sage.modules sage.rings.finite_rings
     1 long time sage.symbolic
     1 sage_numerical_backends_coin
     1 cplex
     1 gurobi
     1 database_knotinfo sage.libs.homfly
     1 snappy
sage-fixdoctests: No fixes made in 'src/sage/knots/all.py'
sage-fixdoctests: No fixes made in 'src/sage/knots/gauss_code.py'
File tags: 
       sage.graphs
Doctest blocks by persistent tags: 
     9 (untagged)
Doctest examples by tags: 
    18 (untagged)
sage-fixdoctests: No fixes made in 'src/sage/knots/knotinfo.py'
File tags: 
       sage.graphs
Doctest blocks by persistent tags: 
    94 (untagged)
     9 sage.groups
     8 sage.modules
     2 database_knotinfo
     1 snappy
     1 not tested
     1 sage.symbolic
Doctest examples by tags: 
   296 (untagged)
    30 database_knotinfo
    21 sage.groups
    14 sage.symbolic
    10 not tested
     9 snappy
sage-fixdoctests: No fixes made in 'src/sage/knots/knot.py'
File tags: 
       sage.graphs
Doctest blocks by persistent tags: 
    19 (untagged)
    16 sage.groups
Doctest examples by tags: 
    91 (untagged)
     4 sage.groups
     1 long time
sage-fixdoctests: No fixes made in 'src/sage/knots/knot_table.py'
```

**Comment by kwankyu** [comment]
> ... After all, in this PR here, there is nothing that actually defines a distribution.

PR #37901 adds `# sage_setup: distribution` directives, but does not actually define distributions. It adds `# sage_setup: distribution = sagemath-graphs` to the files in `sage.knots`.

I am positive to #37901 (and want it to be merged as soon as possible), but I am not to this PR. Let me explain this apparent inconsistency in my position. Adding  `# sage_setup: distribution` directive to files adds almost no burden to others. But 122 `# needs sage.groups` tags to doctests in `sage.knots` adds nontrivial  burden to developers and  users. This burden feels more heavy to me as I think it is due to bad design of `sagemath-graphs`.

> I've added file-level doctest tags `# needs sage.graphs`.

If instead of 122 individual tags a few file-level `# needs sage.groups` tags are added, I can be positive to this PR for now.

Otherwise, let me leave reviewing this PR to others.

**Comment by mkoeppe** [comment]
I've made the suggested change for `link.py`. This changes the statistics to the following:
```
sage-fixdoctests: No fixes made in 'src/sage/knots/link.py'
File tags: 
       sage.graphs sage.groups
Doctest blocks by persistent tags: 
   139 (untagged)
    12 sage.modules
     5 database_knotinfo
     3 sage.libs.homfly
     2 database_knotinfo snappy
     1 sage.modules sage.rings.number_field
     1 sage.symbolic
     1 sage.libs.pari sage.modules
     1 sage.plot
Doctest examples by tags: 
   603 (untagged)
    11 sage.libs.homfly
    10 sage.plot
     9 sage.modules
     8 sage.symbolic
     3 sage.libs.pari sage.modules
     2 database_knotinfo
     1 sage.modules sage.rings.finite_rings
     1 long time sage.symbolic
     1 sage_numerical_backends_coin
     1 cplex
     1 gurobi
     1 database_knotinfo sage.libs.homfly
     1 snappy
```

**Comment by mkoeppe** [comment]
The downside of using the file-level doctest tag is that 
- it is no longer documented what parts of the functionality uses sage.groups
- the parts that do not need groups are no longer tested when sage.groups is not available

It's a tradeoff. 

If your preference is to change them to file-level tags, I can make this change.

**Comment by mkoeppe** [comment]
```
$ ./sage -fixdoctests --no-test --overwrite --verbose src/sage/knots/{knot.py,knotinfo.py}
sage-fixdoctests: Processing unprocessed files
sage-fixdoctests: No fixes made in 'src/sage/knots/knot.py'
File tags: 
       sage.graphs sage.groups
Doctest blocks by persistent tags: 
    35 (untagged)
Doctest examples by tags: 
    95 (untagged)
     1 long time
sage-fixdoctests: No fixes made in 'src/sage/knots/knotinfo.py'
File tags: 
       sage.graphs sage.groups
Doctest blocks by persistent tags: 
   111 (untagged)
     2 database_knotinfo
     1 snappy
     1 not tested
     1 sage.symbolic
Doctest examples by tags: 
   317 (untagged)
    30 database_knotinfo
    14 sage.symbolic
    10 not tested
     9 snappy
```

**Comment by soehms** [comment]
> The downside of using the file-level doctest tag is that
> 
> * it is no longer documented what parts of the functionality uses sage.groups
> * the parts that do not need groups are no longer tested when sage.groups is not available
> 
> It's a tradeoff.
> 
> If your preference is to change them to file-level tags, I can make this change.

I think your second point won't be a problem as long as there are CI runs testing the entire Sage library. 

As for your first point, we lose the work you've already done for this PR. On the other hand, we'll need more work in the future to keep this information valid. 

So the question is, how useful is the information we lose? 

According to your statistics, there are many more untagged doctests than those tagged with `sage.groups`. This looks as if we single out side-cases. However, I doubt there are many non-trivial applications of these modules that don't need `sage.groups` or `sage.modules`. So I prefer the file-level tags.

**Comment by mkoeppe** [comment]
This sounds reasonable to me, thanks.

**Comment by kwankyu** [comment]
```diff
diff --git a/src/sage/knots/knotinfo.py b/src/sage/knots/knotinfo.py
index 6bc9c3a09cf..7942a7092a2 100644
--- a/src/sage/knots/knotinfo.py
+++ b/src/sage/knots/knotinfo.py
@@ -1521,7 +1521,7 @@ class KnotInfoBase(Enum):
 
             sage: K3_1  = KnotInfo.K3_1
             sage: K3_1j = K3_1.jones_polynomial()                                       # needs sage.symbolic
-            sage: L2a1_1j = Ljt     # see note above
+            sage: L2a1_1j = Ljt     # see note above                                    # needs sage.symbolic
             sage: R = L2a1_1j.parent()
             sage: Oj = R(1)
             sage: t = R('t')
```
from
```sage
File "src/sage/knots/knotinfo.py", line 1524, in sage.knots.knotinfo.KnotInfoBase.jones_polynomial
Warning: Variable 'Ljt' referenced here was set only in doctest marked '# needs sage.graphs sage.groups sage.symbolic'
    L2a1_1j = Ljt     # see note above
```

**Comment by kwankyu** [comment]
In
```
$ sage --fixdoctests --verbose src/sage/knots/knot.py 
Running "sage -t -p src/sage/knots/knot.py"
sage-fixdoctests: Processing unprocessed files
sage-fixdoctests: No fixes made in 'src/sage/knots/knot.py'
File tags: 
       sage.graphs sage.groups
Doctest blocks by persistent tags: 
    35 (untagged)
Doctest examples by tags: 
    95 (untagged)
     1 long time
```
`Doctest lines` (parallel to `Doctest blocks`) would make more sense than `Doctest examples`. No?

**Comment by kwankyu** [comment]
Otherwise lgtm. Thanks for switching to file tags.

**Review by kwankyu — APPROVED** [review]
Thanks. LGTM.

**Comment by mkoeppe** [comment]
> `Doctest lines` (parallel to `Doctest blocks`) would make more sense than `Doctest examples`. No?

"Examples" is what the doctest code calls them.

**Comment by kwankyu** [comment]
All right. But I think "example" is ambiguous as a unit of measure...

**Pull request merged**